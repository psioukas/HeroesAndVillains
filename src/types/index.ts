import {z} from "zod";

export interface IHeroType {
    id: string;
    name: string;
}
export const HeroSchema = z.object({
    id: z.string(),
    fullName: z.string(),
    avatarUrl: z.string(),
    description: z.string(),
    type: z.object({
        id: z.string(),
        name: z.string(),
    }),
})

export type IHero = z.infer<typeof HeroSchema>;

export type CreateHeroType = Omit<IHero, 'type' | 'id'> & {
    typeId: string;
};

export interface IHeroesListResponse {
    data: IHero[];
    totalCount: number;
}
