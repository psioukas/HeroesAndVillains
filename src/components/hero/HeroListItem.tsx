import { Paper, PaperProps, Typography } from '@mui/material';
import { styled, useTheme } from '@mui/material/styles';
import { observer } from 'mobx-react-lite';
import React, { useEffect, useRef, useState } from 'react';
import Store from '../../store';
import { IHero } from '../../types';
import HeroAvatar from './HeroAvatar';
const StyledHeroListItem = styled(Paper)<PaperProps>(({ theme }) => ({
    background: theme.palette.common.white,
    paddingLeft: theme.spacing(4),
    paddingRight: theme.spacing(9),
    [theme.breakpoints.up('sm')]: {
        display: 'flex',
        gap: theme.spacing(4),
        flexDirection: 'row',
        justifyContent: 'flex-start',
        alignItems: 'center',
        minHeight: theme.spacing(17.5),
        height: theme.spacing(17.5),
    },
    [theme.breakpoints.down('sm')]: {
        display: 'grid',
        gridTemplateAreas: `
        "hero-avatar hero-name"
        "hero-avatar hero-type"
        "hero-description hero-description"
        `,
        paddingBlock: theme.spacing(4),
        columnGap: theme.spacing(6),
        gridTemplateColumns: `${theme.spacing(11.5)} 1fr`,
        gridTemplateRows: `${theme.spacing(5.75)} ${theme.spacing(5.75)} 1fr`,
        minHeight: theme.spacing(27),
        height: theme.spacing(27),
        '&>:last-child': {
            marginTop: theme.spacing(2),
        },
    },
}));

const HeroListItem: React.FC<{ hero: IHero }> = ({ hero }) => {
    const theme = useTheme();
    const listItemRef = useRef<HTMLDivElement>(null);

    const [mouseOver, setMouseOver] = useState<boolean>(false);
    const handleHoverStart = () => {
        setMouseOver(true);
    };
    const handleHoverFinish = () => {
        setMouseOver(false);
    };
    useEffect(() => {
        if (Store.selectedHeroId === hero.id) {
            Store.setSelectedHero('');
            if (listItemRef.current) {
                listItemRef.current.scrollIntoView({
                    behavior: 'smooth',
                    inline: 'center',
                    block: 'center',
                });

            }
        }
    }, [hero]);
    return (
        <StyledHeroListItem
            id={`hero-list-item-${hero.id}`}
            ref={listItemRef}
            elevation={mouseOver ? 3 : 0}
            onMouseEnter={handleHoverStart}
            onMouseLeave={handleHoverFinish}
            onTouchStart={handleHoverStart}
            onTouchEnd={handleHoverFinish}
            onClick={() => {
                Store.showHeroDetails(hero.id);
            }}
        >
            <HeroAvatar
                src={hero.avatarUrl}
                alt={`${hero.fullName} avatar`}
                height={theme.spacing(11.25)}
                width={theme.spacing(11.25)}
                mr={5}
                gridArea={'hero-avatar'}
            />
            <Typography
                variant={'body1'}
                fontWeight={'bolder'}
                flex={4}
                whiteSpace={'nowrap'}
                overflow={'hidden'}
                textOverflow={'ellipsis'}
                gridArea={'hero-name'}
            >
                {hero.fullName}
            </Typography>
            <Typography
                variant={'body1'}
                flex={3}
                whiteSpace={'nowrap'}
                overflow={'hidden'}
                textOverflow={'ellipsis'}
                gridArea={'hero-type'}
            >
                {hero.type.name}
            </Typography>
            <Typography
                variant={'body1'}
                flex={4}
                whiteSpace={'nowrap'}
                overflow={'hidden'}
                textOverflow={'ellipsis'}
                gridArea={'hero-description'}
            >
                {hero.description}
            </Typography>
        </StyledHeroListItem>
    );
};

export default observer(HeroListItem);
