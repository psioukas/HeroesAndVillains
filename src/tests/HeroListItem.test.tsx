import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import HeroListItem from '../components/hero/HeroListItem';
import Store from '../store/index';
import { IHero } from '../types';
window.HTMLElement.prototype.scrollIntoView = jest.fn();

const hero: IHero = {
    id: 'whatever-id',
    avatarUrl: 'http://www.google.com/images/wtv',
    fullName: 'Full Metal',
    type: {
        id: 'id-human',
        name: 'human',
    },
    description: 'dangerous',
};

const mockedFetch = jest.fn(() =>
    Promise.resolve({
        json: () => Promise.resolve({}),
    }),
) as jest.Mock;

global.fetch = mockedFetch;

test('Renders Hero List Item', () => {
    Store.addHero(hero);
    Store.setSelectedHero(hero.id);
    const domTree = renderer.create(<HeroListItem hero={hero} />).toJSON();
    expect(domTree).toMatchSnapshot();
});

test('Renders Hero List Item find elements', () => {
    render(<HeroListItem hero={hero} />);
    const fullNameText = screen.getByText(hero.fullName);
    const descriptionText = screen.getByText(hero.description);

    expect(fullNameText).toBeInTheDocument();
    expect(descriptionText).toBeInTheDocument();
});
