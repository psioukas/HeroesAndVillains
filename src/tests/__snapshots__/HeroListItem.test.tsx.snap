// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Hero List Item 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 css-exxhi9-MuiPaper-root"
  id="hero-list-item-whatever-id"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onTouchEnd={[Function]}
  onTouchStart={[Function]}
>
  <div
    className="MuiBox-root css-umn51s"
  >
    <img
      alt="Full Metal avatar"
      height="90px"
      src="http://www.google.com/images/wtv"
      width="90px"
    />
  </div>
  <p
    className="MuiTypography-root MuiTypography-body1 css-1slgovd-MuiTypography-root"
  >
    Full Metal
  </p>
  <p
    className="MuiTypography-root MuiTypography-body1 css-1748e8y-MuiTypography-root"
  >
    human
  </p>
  <p
    className="MuiTypography-root MuiTypography-body1 css-7fl3xu-MuiTypography-root"
  >
    dangerous
  </p>
</div>
`;
